File: serverCode/store.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime, timezone


class Record(BaseModel):
    '''
        A record of the sensor data
        datetime: The time the record was created (UTC)
        battery: The battery voltage
        temperature: The temperature in °C
        humidity: The humidity in %
        moisture: The moisture in %
    '''
    time: str = Field(default_factory=lambda: datetime.utcnow().astimezone(timezone.utc).isoformat())
    battery: float | None = None
    temperature:  float | None = None
    humidity: float | None = None
    moisture: float | None = None

    def to_csv(self) -> str:
        '''
            Returns a csv string of the record
        '''
        return ",".join(
            [
                str(self.time) if self.time is not None else "",
                str(self.battery) if self.battery is not None else "",
                str(self.temperature) if self.temperature is not None else "",
                str(self.humidity) if self.humidity is not None else "",
                str(self.moisture) if self.moisture is not None else "",
            ]
        )

    @staticmethod
    def from_csv(csv_entry : str):
        '''
            Creates a record from a csv string
            cvc_entry: The csv string
        '''
        dt, battery, temperature, humidity, moisture = csv_entry.replace('\n', '').split(",")
        if dt == "":
            raise ValueError(f"Missing datetime in record {csv_entry}")
        return Record(
            time= str(datetime.fromisoformat(dt)),
            battery= float(battery) if battery != "" else None,
            temperature= float(temperature) if temperature != "" else None,
            humidity= float(humidity) if humidity != "" else None,
            moisture= float(moisture) if moisture != "" else None,
        )
    def __ge__(self, other : datetime):
        return datetime.fromisoformat(self.time) >= other
    
    def __le__(self, other : datetime):
        return datetime.fromisoformat(self.time) <= other

class Store:
    '''
        A store for the records
        path: The path to the csv file
    '''
    def __init__(self, path):
        self.path = path
        try:
            with open(self.path, "r") as f:
                self.records = [Record.from_csv(line) for line in f.readlines()]
        except FileNotFoundError:
            self.records = []

    def write(self, record: Record):
        '''
            Writes a record to the store
            record: The record to write
        '''
        self.records.append(record)

    def save(self):
        '''
            Saves the store to the csv file
        '''
        with open(self.path, "w") as f:
            for record in self.records:
                f.write(record.to_csv() + "\n")

    def read(self, last_n : Optional[int] = None):
        '''
            Reads the records from the store
            from_time: The time from which to read the records
            to_time: The time to which to read the records
        '''
        return self.records[-last_n:] if last_n is not None else self.records
        

File: serverCode/src/store.rs
use chrono::{DateTime, TimeZone, Utc};
use csv;
use derive_getters::Getters;
use serde::{Deserialize, Serialize};
use std::{error, path, result};

type Result<T> = result::Result<T, Box<dyn error::Error>>;

macro_rules! filter_time {
    ($option : ident, $op : tt) => {
        |r : &Record|{
            if let Some(time) = $option {
                if !(r.time $op *time) {
                    return false;
                }
            }
            true
        }
    };
}

#[derive(Debug, Serialize, Deserialize, Clone, Getters, PartialEq)]
pub struct Record {
    #[serde(default="Utc::now")]
    time: DateTime<Utc>,
    battery: Option<f32>,
    temperature: Option<f32>,
    humidity: Option<f32>,
    moisture: Option<f32>,
}

impl Record {
    pub fn new_without_time(
        battery: Option<f32>,
        temperature: Option<f32>,
        humidity: Option<f32>,
        moisture: Option<f32>,
    ) -> Self {
        Record {
            time: Utc::now(),
            battery,
            temperature,
            humidity,
            moisture,
        }
    }
    pub fn new<Tz: TimeZone>(
        time: DateTime<Tz>,
        battery: Option<f32>,
        temperature: Option<f32>,
        humidity: Option<f32>,
        moisture: Option<f32>,
    ) -> Self {
        Record {
            time: time.with_timezone(&Utc),
            battery,
            temperature,
            humidity,
            moisture,
        }
    }
}

#[derive(Debug)]
pub struct Store {
    path: String,
    records: Vec<Record>,
}

impl Store {
    pub fn open(path: &str) -> Result<Self> {
        let records = Self::load_from_file(path)?;
        Ok(Store {
            path: path.to_string(),
            records,
        })
    }

    fn load_from_file(path: &str) -> Result<Vec<Record>> {
        let mut records = Vec::new();
        let path = path::Path::new(path);
        if path.exists() {
            let mut file = csv::Reader::from_path(path)?;
            for record in file.deserialize() {
                let record = record?;
                records.push(record);
            }
        }
        Ok(records)
    }

    pub fn write(&mut self, record: Record) {
        self.records.push(record);
    }

    pub fn path(&self) -> &str {
        &self.path
    }

    pub fn save(&self) -> Result<()> {
        let mut writer = csv::Writer::from_path(&self.path)?;
        for record in &self.records {
            writer.serialize(record)?;
        }
        writer.flush()?;
        Ok(())
    }

    pub fn read(
        &self,
        start_time: &Option<DateTime<Utc>>,
        end_time: &Option<DateTime<Utc>>,
    ) -> Vec<Record> {
        self.records
            .iter()
            .filter(|r| filter_time!(start_time, >=)(r) && filter_time!(end_time, <=)(r))
            .map(|r| r.to_owned())
            .collect()
    }
    pub fn erase(&mut self, start_time: &Option<DateTime<Utc>>, end_time: &Option<DateTime<Utc>>) {
        self.records
            .retain(|r| !filter_time!(start_time, >=)(r) || !filter_time!(end_time, <=)(r))
    }
}
